src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.DocumentChangeTracker.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    internal partial class RequestExecutionQueue
+    {
+        /// <summary>
+        /// Keeps track of changes to documents that are opened in the LSP client. Calls MUST not overlap, so this
+        /// should be called from a mutating request handler. See <see cref="RequestExecutionQueue"/> for more details.
+        /// </summary>
+        internal class DocumentChangeTracker : IWorkspaceService
+        {
+            private readonly Dictionary<(Workspace, DocumentId), Document> _trackedDocuments = new();
```

Comment:
Presumably with this change, the RequestExecutionQueue would be updating the source text of the document, rather than replacing the document in the solution, in order to give the request the latest copy, right? Would that mean that every request causes a re-parse of the syntax tree for any document it accesses?

Reply:
Talked offline, updated these to SourceText